std/algorithm/comparison.d

```diff
@@ -1565,3 +1568,228 @@ unittest
                 2, "two",
                 ));
 }
+
+/**
+Returns true if all of the elements in $(D r1) are present in $(D r2). Otherwise,
+returns false.
+
+If your parameters are forward ranges, then this function can allocate if the
+$(D AllocateGC.yes) flag is passed. This has the benefit of being much
+faster than the $(D AllocateGC.no) option.
+
+Or, if your two ranges are sortable, then passing sorted ranges will be faster
+than the other two options. If your sorted ranges need custom equality, you
+can override the default $(D "a == b") predicate. See $(D binaryFun) for more
+details
+
+Non allocating forward range option: $(BIGOH r1.length * r2.length)
+Allocating forward range option: $(BIGOH r1.length) + $(BIGOH r2.length)
+Non allocating sorted range option: $(BIGOH min(r1.length, r2.length))
+
+Params:
+    r1 = Any finite forward range
+    r2 = Any finite forward range
+
+Returns:
+    $(D bool)
+*/
+
+alias AllocateGC = Flag!"AllocateGC";
+
+bool isPermutation(Range1, Range2)(Range1 r1, Range2 r2, AllocateGC allocate_gc = AllocateGC.no)
+    if (isForwardRange!Range1 &&
+        isForwardRange!Range2 &&
+        !isInfinite!Range1 &&
+        !isInfinite!Range2)
+{
```

Comment:
The allocation option simply *does not work* for any predicate other than true equality. So the predicate shouldn't even be specifiable in that case. A function should not take a parameter that it cannot implement. The `allocate_gc` parameter should be compile-time rather than runtime, and there should be two overloads:
````
bool isPermutation(AllocateGC allocGC, R1, R2)(R1 r1, R2 r2)
    if (allocateGC == AllocateGC.yes)
{
    ....
}

bool isPermutation(AllocateGC allocGC, alias pred = `a==b`, R1, R2)(R1 r1, R2 r2)
    if (allocateGC == AllocateGC.no)
{
    ...
}
````
Or something along these lines.